### PR TITLE
New Json ticket validator to accept CAS responses in JSON.

### LINKED ISSUE #211: Design a new validator to accept a JSON-formatted response
This task proposes a small enhancement to the client where a new validator is added to accept/parse the CAS validation response that may be produced as JSON. Ideally, if the JSON parsing fails, the validator may fall back onto XML. 

Producing a CAS response in JSON at times is helpful specially when it comes to encoding attributes that are named weird, such as "this:is:not:acceptable:in:xml". 

Examples include:

```json
{
  "serviceResponse" : {
    "authenticationSuccess" : {
      "user" : "username",
      "proxyGrantingTicket" : "PGTIOU-84678-8a9d..."
    }
  }
}
```

Or

```json
{
  "serviceResponse" : {
    "authenticationFailure" : {
      "code" : "INVALID_TICKET",
      "description" : "Ticket ST-1856339-aA5Yuvrxzpv8Tau1cYQ7 not recognized"
    }
  }
}
```

Or

```json
{
  "serviceResponse" : {
    "authenticationSuccess" : {
      "user" : "username",
      "proxyGrantingTicket" : "PGTIOU-84678-8a9d...",
      "proxies" : [ "https://proxy1/pgtUrl", "https://proxy2/pgtUrl" ],
      "attributes" : {
        "firstName" : "John",
        "affiliation" : [ "staff", "faculty" ],
        "title" : "Mr.",
        "email" : "jdoe@example.orgmailto:jdoe@example.org",
        "lastname" : "Doe"
      }
    }
  }
}
```

I believe the code already exists in cas-addons somewhere to handle the client parsing.

### PR DESCRIPTION
**Pull request by mmoayyed**
Closes #211 

This pull requests follows a similar model to other CAS filters to add a `Cas30JsonProxyReceivingTicketValidationFilter`.  The filter by default relies upon `Cas30JsonServiceTicketValidator` and `Cas30JsonProxyTicketValidator` to  handle JSON responses for all cases, and also attempts to fallback onto the current XML response if needed. The latter behavior is needed to handle cases where the validator asks for JSON and the server for whatever reason decides to ignore that request and produce an XML response instead. 

The readme is updated with additional info. The JSON snippets that are worked into the validators are based on what the CAS protocol recommends, and there are accompanying test cases to ensure correctness. 

The pull uses jackson-databind to handle JSON parsing, and I also had to fix a few atlassian-related dependencies in order to ensure the build can build and compile locally. There still seem to be a few outdated dependencies and cause travis to fail and I'll work through those later, but in the meantime, this should be a good starting point for review.

### CODE REVIEW

**Comment by mmoayyed** [comment]
This is ready to be reviewed.

**Comment by mmoayyed** [comment]
Gentle ping that this PR is ready to be reviewed.

**Comment by serac** [comment]
I'll try to give some review cycles tonight.

File: README.md [hunk 1]
```diff
@@ -342,7 +342,14 @@ Validates the tickets using the CAS 2.0 protocol. If you provide either the `acc
...
-Validates the tickets using the CAS 3.0 protocol. If you provide either the `acceptAnyProxy` or the `allowedProxyChains` parameters, a `Cas30ProxyTicketValidator` will be constructed. Otherwise a general `Cas30ServiceTicketValidator` will be constructed that does not accept proxy tickets. Supports all configurations that are available for `Cas20ProxyReceivingTicketValidationFilter`.
+Validates the tickets using the CAS 3.0 protocol. If you provide either the `acceptAnyProxy` or the `allowedProxyChains` parameters, 
+a `Cas30ProxyTicketValidator` will be constructed. Otherwise a general `Cas30ServiceTicketValidator` will be constructed that does not 
+accept proxy tickets. Supports all configurations that are available for `Cas20ProxyReceivingTicketValidationFilter`.
+
+#### org.jasig.cas.client.validation.json.Cas30JsonProxyReceivingTicketValidationFilter
+Indentical to `Cas30ProxyReceivingTicketValidationFilter`, yet the filter is able to accept validation responses from CAS
+that are formatted as JSON per guidelines laid out by the CAS protocol. See the [protocol documentation](https://apereo.github.io/cas/)
```
Comment (serac) [inline README.md]: Linking directly to the protocol document, https://apereo.github.io/cas/5.1.x/protocol/CAS-Protocol-Specification.html,  would be helpful.

File: cas-client-core/src/main/java/org/jasig/cas/client/validation/json/Cas30JsonServiceTicketValidator.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        super(casServerUrlPrefix);
+        setCustomParameters(Collections.singletonMap("format", "JSON"));
+    }
+
+    @Override
+    protected Assertion parseResponseFromServer(final String response) throws TicketValidationException {
+        try {
+            final TicketValidationJsonResponse json = new JsonValidationResponseParser().parse(response);
+            return json.getAssertion(getProxyGrantingTicketStorage(), getProxyRetriever());
+        } catch (final Exception e) {
```
Comment (serac) [inline cas-client-core/src/main/java/org/jasig/cas/client/validation/json/Cas30JsonServiceTicketValidator.java]: I like the idea of falling back to XML parsing on JSON parse errors, but switching on `Exception` seems too broad. Consider `com.fasterxml.jackson.core.JsonProcessingException`, which appears to cover low-level parse errors (`JsonParseException `) as well as JSON content that doesn't map to the `TicketValidationJsonResponse` model (`JsonMappingException`).

File: cas-client-core/src/main/java/org/jasig/cas/client/validation/json/TicketValidationJsonResponse.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            assertion = new AssertionImpl(attributePrincipal);
+        } else {
+            assertion = new AssertionImpl(new AttributePrincipalImpl(principal, attributes));
+        }
+        return assertion;
+    }
+
+    static class CasServiceResponseAuthentication {
+        private CasServiceResponseAuthenticationFailure authenticationFailure;
+        private CasServiceResponseAuthenticationSuccess authenticationSuccess;
```
Comment (serac) [inline cas-client-core/src/main/java/org/jasig/cas/client/validation/json/TicketValidationJsonResponse.java]: It strikes me as odd that there are slots for both success and failure, but if that's what the protocol specifies I suppose it's necessary. I recall seeing checks in the validator that both fields are not set simultaneously, which is the concern in practice, so I suppose it's all good. If the JSON parser didn't impose requirements on the JSON structure, making the fields final and setting them in the constructor would be better.

**Comment by mmoayyed** [comment]
Outstanding. Thank you. All changes applied.

File: cas-client-core/src/main/java/org/jasig/cas/client/validation/json/TicketValidationJsonResponse.java (hunk 3 above)
Comment (mmoayyed) [inline cas-client-core/src/main/java/org/jasig/cas/client/validation/json/TicketValidationJsonResponse.java]: All done. A quick note that the protocol of course requires that only of those slots be returned. It,  the response, is either a success or failure, both of which at a given time are encapsulated inside a "serviceResponse" parent "tag". The two-slot situation here with the fields is mostly weirdness introduced by jackson parsing, and in a way, it's a bit of a shortcut. Impl detail, as it were.

**Comment by mmoayyed** [comment]
@battags @leleuj Do you think you might have time to review this patch also before the weekend?

File: cas-client-core/src/main/java/org/jasig/cas/client/validation/json/Cas30JsonProxyTicketValidator.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+public class Cas30JsonProxyTicketValidator extends Cas30ProxyTicketValidator {
+    public Cas30JsonProxyTicketValidator(final String casServerUrlPrefix) {
+        super(casServerUrlPrefix);
+        setCustomParameters(Collections.singletonMap("format", "JSON"));
+    }
+
+    @Override
+    protected Assertion parseResponseFromServer(final String response) throws TicketValidationException {
+        try {
+            final TicketValidationJsonResponse json = new JsonValidationResponseParser().parse(response);
```
Comment (leleuj) [inline cas-client-core/src/main/java/org/jasig/cas/client/validation/json/Cas30JsonProxyTicketValidator.java]: Cannot we share this `JsonValidationResponseParser`?

**Comment by leleuj** [comment]
Looks good to me with one comment

File: cas-client-core/src/main/java/org/jasig/cas/client/validation/json/Cas30JsonProxyTicketValidator.java (hunk 4 above)
Comment (mmoayyed) [inline cas-client-core/src/main/java/org/jasig/cas/client/validation/json/Cas30JsonProxyTicketValidator.java]: Share in what way? You mean from a pac4j perspective perhaps outside the core client project? 

You shouldn't need it. The validator is all that you would care about. I did start off trying to get the parser into the validator as an injection point, but the hierarchy of validators and such made it difficult to not duplicate code/logic. Can share previous examples of what I did, if that helps.

**Pull request merged**